zstd-decompress.c

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2024, Stefan Hansson
+ */
+
+/* For memfd_create */
+#define _GNU_SOURCE
+
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <zstd.h>
+
+#include "zstd-decompress.h"
+
+/**
+ * zstd_decompress_file() - decompress a zstd-compressed file
+ * @filename:	path to a file to decompress
+ *
+ * Return: opened fd on success, -1 on error
+ */
+int zstd_decompress_file(const char *filename)
+{
+	/* Figure out the size of the file. */
+	struct stat file_stat;
+	if (stat(filename, &file_stat) == -1) {
+		perror("stat failed");
+		return -1;
+	}
+
+	const off_t file_size = file_stat.st_size;
+	const size_t size_t_file_size = (size_t)file_size;
+	/* Set up buffer, read file, then decompress. */
+	void* const compressed_buffer = malloc(size_t_file_size);
+	if (compressed_buffer == NULL) {
+		perror("malloc failed");
+		return -1;
+	}
+
+	FILE* const input_file = fopen(filename, "rb");
+	if (input_file == NULL) {
+		perror("fopen failed");
+		return -1;
+	}
+
+	fread(compressed_buffer, 1, size_t_file_size, input_file);
+	if (ferror(input_file)) {
+		perror("fread failed");
+		return -1;
+	}
+
+	const unsigned long long decompressed_size = ZSTD_getFrameContentSize(compressed_buffer, size_t_file_size);
+	if (decompressed_size == ZSTD_CONTESigned-off-by: Nikita Travkin <nikita@trvn.ru>NTSIZE_UNKNOWN) {
```

Comment:
Ah, Sign-off is not missing after all! :D

Reply:
Oops, not sure how that happened 🙃